src/compiler/scala/tools/nsc/transform/patmat/Logic.scala

```diff
@@ -174,13 +174,55 @@ trait Logic extends Debugging  {
      */
     def simplify(f: Prop): Prop = {
 
-      // limit size to avoid blow up
-      def hasImpureAtom(ops: Seq[Prop]): Boolean = ops.size < 10 &&
-        ops.combinations(2).exists {
-          case Seq(a, Not(b)) if a == b => true
-          case Seq(Not(a), b) if a == b => true
-          case _                        => false
+      def hasImpureAtom(ops0: collection.Iterable[Prop]): Boolean = {
+        // HOT method, imperative rewrite of:
+        // ops.combinations(2).exists {
+        //   case Seq(a, Not(b)) if a == b => true
+        //   case Seq(Not(a), b) if a == b => true
+        //   case _                        => false
+        // }
+
+        def checkPair(a: Prop, b: Prop): Boolean = {
+          b match {
+            case Not(b) if a == b => true
+            case _ =>
+              a match {
+                case Not(a) if a == b => true
+                case _ => false
+              }
+          }
+        }
+        val size = ops0.size
```

Comment:
Is collection.Iterable.size guaranteed to be O(1)?

Reply:
We pass in an `immutable.HashSet` here, which has constant time `size`. I could sharpen up the type, but its just a local method.

Reply:
Got it. It's not likely to matter here, but in general having a concrete type to make the invocation against is better to avoid those costly invokeinterface's